pkg/controller/admissionchecks/multikueue/multikueuecluster.go

```diff
@@ -0,0 +1,339 @@
+		}
+	}()
+
+	rc.watchCancel = newWatcher.Stop
+	rc.kubeconfig = kubeconfig
+	return nil
+}
+
+func (rc *remoteClient) queueWorkloadEvent(ctx context.Context, ev watch.Event) {
+	wl, isWl := ev.Object.(*kueue.Workload)
+	if !isWl {
+		return
+	}
+
+	localWl := &kueue.Workload{}
+	if err := rc.localClient.Get(ctx, client.ObjectKeyFromObject(wl), localWl); err == nil {
+		rc.wlUpdateCh <- event.GenericEvent{Object: localWl}
+	}
+}
+
+// clustersReconciler implements the reconciler for all MultiKueueClusters.
+// Its main task being to maintain the list of remote clients associated to each MultiKueueCluster.
+type clustersReconciler struct {
+	client          client.Client
+	configNamespace string
+
+	lock sync.RWMutex
+	// The list of remote clients, indexed by the cluster name.
+	clients    map[string]*remoteClient
+	wlUpdateCh chan event.GenericEvent
+
+	// rootContext - holds the context passed by the controller-runtime on Start.
+	// It's used to create child contexts for MultiKueueClusters client watch routines
+	// that will gracefully end when the controller-manager stops.
+	rootContext context.Context
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	builderOverride clientWithWatchBuilder
+}
+
+var _ manager.Runnable = (*clustersReconciler)(nil)
+var _ reconcile.Reconciler = (*clustersReconciler)(nil)
+
+func (c *clustersReconciler) Start(ctx context.Context) error {
+	c.rootContext = ctx
+	return nil
+}
+
+func (c *clustersReconciler) stopAndRemoveCluster(clusterName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if rc, found := c.clients[clusterName]; found {
+		rc.watchCancel()
+		delete(c.clients, clusterName)
+	}
+}
+
+func (c *clustersReconciler) setRemoteClientConfig(clusterName string, kubeconfig []byte) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	client, found := c.clients[clusterName]
+	if !found {
+		client = newRemoteClient(c.client, c.wlUpdateCh)
+		if c.builderOverride != nil {
+			client.builderOverride = c.builderOverride
+		}
+		c.clients[clusterName] = client
+	}
+
+	if err := client.setConfig(c.rootContext, kubeconfig); err != nil {
+		delete(c.clients, clusterName)
+		return err
+	}
+	return nil
+}
+
+func (a *clustersReconciler) controllerFor(acName string) (*remoteClient, bool) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	c, f := a.clients[acName]
+	return c, f
+}
+
+func (c *clustersReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	cluster := &kueuealpha.MultiKueueCluster{}
+
+	err := c.client.Get(ctx, req.NamespacedName, cluster)
+	if client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err != nil || !cluster.DeletionTimestamp.IsZero() {
+		c.stopAndRemoveCluster(req.Name)
+		// notify ac reconcilere
```

Comment:
Is this accurate? IIUC it just re-triggers reconciliation of the `clustersReconciler`.

Reply:
No longer, the admission check reconciler watches the clusters and will get the event from the status update .

removed

Reply:
ok